py4DSTEM/process/diffraction/crystal_phase.py

```diff
@@ -33,306 +35,1366 @@ def __init__(
+                        basis[:, inds_solve],
+                        obs,
+                    )
+
+                    if (
+                        np.count_nonzero(phase_weights_cand > 0.0)
+                        <= max_number_patterns
+                    ):
+                        phase_weights[inds_solve] = phase_weights_cand
+                        phase_residual = phase_residual_cand
+                        search = False
+                    else:
+                        inds = np.where(inds_solve)[0]
+                        inds_solve[inds[np.argmin(phase_weights_cand)]] = False
+
+                # Estimate the phase reliability
+                inds_solve = np.ones(self.num_fits, dtype="bool")
+                inds_solve[phase_weights > 1e-8] = False
+
+                if np.all(inds_solve == False):
+                    phase_reliability = 0.0
+                else:
+                    search = True
+                    while search is True:
+                        phase_weights_cand, phase_residual_cand = nnls(
+                            basis[:, inds_solve],
+                            obs,
+                        )
+                        if (
+                            np.count_nonzero(phase_weights_cand > 0.0)
+                            <= max_number_patterns
+                        ):
+                            phase_residual_2nd = phase_residual_cand
+                            search = False
+                        else:
+                            inds = np.where(inds_solve)[0]
+                            inds_solve[inds[np.argmin(phase_weights_cand)]] = False
+
+                    phase_weights_cand, phase_residual_cand = nnls(
+                        basis[:, inds_solve],
+                        obs,
+                    )
+                    phase_reliability = phase_residual_2nd - phase_residual
+
+        except:
+            phase_weights = np.zeros(self.num_fits)
+            phase_residual = np.sqrt(np.sum(intensity**2))
+            phase_reliability = 0.0
+
+        if verbose:
+            ind_max = np.argmax(phase_weights)
+            # print()
+            print("\033[1m" + "phase_weight   or_ind   name" + "\033[0m")
+            # print()
+            for a0 in range(self.num_fits):
+                c = self.crystal_identity[a0, 0]
+                m = self.crystal_identity[a0, 1]
+                line = "{:>12} {:>8}   {:<12}".format(
+                    f"{phase_weights[a0]:.2f}", m, self.crystal_names[c]
+                )
+                if a0 == ind_max:
+                    print("\033[1m" + line + "\033[0m")
+                else:
+                    print(line)
+            print("----------------------------")
+            line = "{:>12} {:>15}".format(f"{sum(phase_weights):.2f}", "fit total")
+            print("\033[1m" + line + "\033[0m")
+            line = "{:>12} {:>15}".format(f"{phase_residual:.2f}", "fit residual")
+            print(line)
```

Comment:
This isn't necessary to change, but I'd just point out the `tabulate` package, which is excellent for this kind of thing: https://github.com/astanin/python-tabulate

Reply:
This was very tedious to implement, next time I will def check out this package!